clang/lib/Sema/SemaModule.cpp

```diff
@@ -924,6 +939,23 @@ static bool checkExportedDeclContext(Sema &S, DeclContext *DC,
 /// Check that it's valid to export \p D.
 static bool checkExportedDecl(Sema &S, Decl *D, SourceLocation BlockStart) {
 
+  // HLSL: export declaration is valid only on functions
+  if (S.getLangOpts().HLSL) {
+    auto *FD = dyn_cast<FunctionDecl>(D);
+    if (!FD) {
+      if (auto *ED2 = dyn_cast<ExportDecl>(D)) {
+        S.Diag(ED2->getBeginLoc(), diag::err_export_within_export);
+        if (auto *ED1 = dyn_cast<ExportDecl>(D->getDeclContext()))
+          S.Diag(ED1->getBeginLoc(), diag::note_export);
+      }
```

Comment:
The export_within_export case looks like it's handled in ActOnStartExportDecl in C++ - is HLSL's behaviour different in that respect?

Reply:
On this note, we should also add a test like so:
```hlsl
export {
  export { // expected-error {{export declaration appears within another export declaration}}
    void f();
  }
}
```

Reply:
Good catch!